crypto/evp/digest.c

```diff
@@ -140,7 +140,26 @@ int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl)
     }
     if (ctx->flags & EVP_MD_CTX_FLAG_NO_INIT)
         return 1;
-    return ctx->digest->init(ctx);
+    if (ctx->digest->init(ctx) == 0)
+        return 0;
+
+    if (ctx->pctx != NULL) {
+        int r;
+
+        ERR_set_mark();
+
+        r = EVP_PKEY_CTX_ctrl(ctx->pctx, -1, EVP_PKEY_OP_TYPE_SIG,
+                              EVP_PKEY_CTRL_DIGESTSETUP, 0, ctx);
+
+        if (r == -2)
+            ERR_pop_to_mark();
+
+        ERR_clear_last_mark();
```

Comment:
`ERR_pop_to_mark()` also clears the last mark, so this call should be part of an `else` in the above `if`.